### ISSUE TITLE
2d electrostatic methods don't agree with each other 

### ISSUE THREAD

**Issue by RudolfWeeber** [espressomd/espresso#1559]
@KonradBreitsprecher @arnolda 

Discovered while writing tests (#1324).
I tested: p3m+elc, mmm2d, scafacos p2nfft against each other and against reference data obtained by direct summation with a circle-cutoff with a radius of 400 periodic images.
* scafacos p2nfft agreed with the direct summation result (Deviation: forces: abs 1.6E-3, abs energy 0.08, rel=1E-3). Energy should converge much slower. I'm generating test data with 2000 images and will update, once finished.
* mmm2d forces match reference data also up to 1.6E-3. The forces seem to agree with p2nfft. Energy is off completely (deviation of abs 55, rel 25%)
* P3m+elc somewhat agrees with reference data and p2nfft, but is an order of magnitude worse even though I tuned p3m to 1E-7 and elc pair wise error to 1E-9. Force difference is 0.05 and energy difference 0.6.

The system is the coulomb_cloud_wall test from scafacos with the gap region from z=9 to z=10=box_l cleared for elc.

In conclusion, I think, something is wrong with elc forces+energy and mmm2d energy. It is not clear to me whether this is in the interface or the core.

Can someone with experience in these methods please take this over?

To reproduce:
The test + reference data is on my github in the branch electrostatic_mixed_periodicity.
Test is in testsuite/coulomb_mixed_periodicity.py. Not included in CMakeList and the assertions are turned off at this point.
I put the reference data and the script to produce it in the testsuite/data folder.

To use the scafacos methods on ICP Ubuntu machines:
export PKG_CONFIG_PATH=/usr/local/stow/scafacos/lib/pkgconfig/scafacos.pc:$PKG_CONFIG_PATH
and 
export LD_LIBRARY_PATH=/usr/local/stow/scafacos/lib/:$LD_LIBRARY_PATH
before cmake.

**Comment by KonradBreitsprecher** [comment]
I found that in the calculation of the self-energy, the sum over local particles changed to a std::accumulate, which is broken:
 
```
 double seng = coulomb.prefactor * calc_mmm2d_copy_pair_energy(dv);
 auto parts = local_cells.particles();
 self_energy = std::accumulate(
     parts.begin(), parts.end(), 0,
     [seng](double sum, Particle const &p) { return sum + seng * SQR(p.p.q); });
```
There is a subtle `0` to initialize the sum, but also sets the type to int. 
Changing this to a double `0.0` corrects the mmm2d energy in Rudolfs testcase.
Note that I also found a similar call of std::accumulate in mdlc_correction.cpp,
which might be broken as well.

**Comment by RudolfWeeber** [comment]
I merged the current Python branch and  #1580.
MMM2D is fixed (results agree with p2nfft for forces and energy). Elc, despite the bugfix has not improved, apparently.

**Comment by KonradBreitsprecher** [comment]
I found another small error in the tuning routine of ELC, but it did not improve the accuracy in the cloud test in coulomb_mixed_periodicity.py.
However, with an increased ELC gap size, ELC produces rms force and energy differences comparable to MMM2D. I tested coulomb_mixed_periodicity.py with a gap size of 3 (by increasing box_l[2] to 12) and got

- mmm2d rms force difference 0.00160918028392
- mmm2d energy difference 0.0818253216829
- elc rms force difference 0.00160910526549
- elc energy difference 0.0815047602653

**Comment by RudolfWeeber** [comment]
Offline discussion result:
* When the gap is small (10% of box), the elc tuning does not result in parameters which give an accuracy comparable to other methods (p2nfft and mmm2d). We will put a wraning in the user guide and the interface for elc. Users should choose larger gaps and compare force and energy to mmm2d for a representative configuration of their system.
* Elc with image charges is likely fixed, @KonradBreitsprecher contributes a test case whcih compares mmm2d and elc with image charges.
* The mmm2d energy is fixed
* ElectrostaticExtensions cannot be disabled safely without disabling the underlying electrostatics method as well. We will prevent their removal from a system by raising exceptions in _deactivate_method().

**Comment by KonradBreitsprecher** [comment]
Can be closed via #1638 . electrostatic_mixed_periodicity @RudolfWeeber is merged.
